File: demo-apps/demoapp001-typescript-cypress/app_src/audit/AuditLogger.ts
import { AuditConfig, AuditEvent, AuditStatistics, AuditTrail, CellChange } from './AuditTypes';

const DEFAULT_CONFIG: AuditConfig = {
  enabled: true,
  includeGridSnapshots: false,
  verbosityLevel: 'standard',
};

export class AuditLogger {
  private readonly config: AuditConfig;
  private readonly events: AuditEvent[] = [];
  private readonly puzzleName: string;
  private readonly initialGrid: number[][];
  private readonly startTime: Date;
  private currentIteration: number = 0;
  private eventIdCounter: number = 0;

  constructor(puzzleName: string, initialGrid: number[][], config: Partial<AuditConfig> = {}) {
    this.puzzleName = puzzleName;
    this.initialGrid = initialGrid.map((r) => [...r]);
    this.startTime = new Date();
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  isEnabled(): boolean {
    return this.config.enabled;
  }

  startIteration(): void {
    this.currentIteration++;
  }

  logChange(
    algorithm: AuditEvent['algorithm'],
    cellChanges: CellChange[],
    algorithmParameter?: number,
    gridSnapshotAfter?: number[][]
  ): void {
    if (!this.config.enabled || cellChanges.length === 0) return;
    const event: AuditEvent = {
      eventId: ++this.eventIdCounter,
      timestamp: new Date().toISOString(),
      iteration: this.currentIteration,
      algorithm,
      cellChanges,
      ...(algorithmParameter !== undefined && { algorithmParameter }),
      ...(this.config.includeGridSnapshots &&
        gridSnapshotAfter && {
          gridSnapshotAfter: gridSnapshotAfter.map((r) => [...r]),
        }),
    };
    this.events.push(event);
  }

  getChangeCount(): number {
    return this.events.reduce((sum, e) => sum + e.cellChanges.length, 0);
  }

  getStatistics(): AuditStatistics {
    const stats: AuditStatistics = {
      changesByAlgorithm: { unitCompletion: 0, hiddenSingles: 0, nakedSingles: 0 },
      iterationsByAlgorithm: { unitCompletion: 0, hiddenSingles: 0, nakedSingles: 0 },
      averageChangesPerIteration: 0,
    };
    for (const event of this.events) {
      const changes = event.cellChanges.length;
      if (event.algorithm === 'UnitCompletion') {
        stats.changesByAlgorithm.unitCompletion += changes;
        stats.iterationsByAlgorithm.unitCompletion++;
      } else if (event.algorithm === 'HiddenSingles') {
        stats.changesByAlgorithm.hiddenSingles += changes;
        stats.iterationsByAlgorithm.hiddenSingles++;
      } else {
        stats.changesByAlgorithm.nakedSingles += changes;
        stats.iterationsByAlgorithm.nakedSingles++;
      }
    }
    const total =
      stats.changesByAlgorithm.unitCompletion +
      stats.changesByAlgorithm.hiddenSingles +
      stats.changesByAlgorithm.nakedSingles;
    stats.averageChangesPerIteration =
      this.currentIteration > 0 ? total / this.currentIteration : 0;
    return stats;
  }

  getTrail(finalGrid: number[][], status: 'SOLVED' | 'STUCK_ON_ADVANCED_LOGIC'): AuditTrail {
    const endTime = new Date();
    return {
      puzzleName: this.puzzleName,
      startTime: this.startTime.toISOString(),
      endTime: endTime.toISOString(),
      totalDurationMs: endTime.getTime() - this.startTime.getTime(),
      initialGrid: this.initialGrid,
      finalGrid: finalGrid.map((r) => [...r]),
      status,
      totalIterations: this.currentIteration,
      totalChanges: this.getChangeCount(),
      events: this.events,
      statistics: this.getStatistics(),
    };
  }
}

File: demo-apps/demoapp003-csharp-specflow/app_src/AuditLogger.cs
namespace DemoApp003.Sudoku;

public sealed class AuditLogger
{
    private readonly string _puzzleName;
    private readonly int[][] _initialGrid;
    private readonly AuditConfig _config;
    private readonly List<AuditEvent> _events = [];
    private int _currentIteration;

    public AuditLogger(string puzzleName, IReadOnlyList<IReadOnlyList<int>> initialGrid, AuditConfig config)
    {
        _puzzleName = puzzleName;
        _initialGrid = GridHelpers.DeepCopy(initialGrid);
        _config = config;
    }

    public bool IsEnabled() => _config.Enabled;

    public void StartIteration() => _currentIteration++;

    public void LogChange(string algorithm, IReadOnlyList<CellChange> changes, int? algorithmParam = null)
    {
        if (!IsEnabled() || changes.Count == 0)
        {
            return;
        }

        _events.Add(new AuditEvent(_currentIteration, algorithm, algorithmParam, changes.ToArray()));
    }

    public AuditTrail GetTrail(IReadOnlyList<IReadOnlyList<int>> finalGrid, string status)
    {
        var totalChanges = _events.Sum(item => item.Changes.Count);
        var statistics = new AuditStatistics(
            UnitCompletion: CountChanges("UnitCompletion"),
            HiddenSingles: CountChanges("HiddenSingles"),
            NakedSingles: CountChanges("NakedSingles"));

        return new AuditTrail(
            _puzzleName,
            _initialGrid,
            GridHelpers.DeepCopy(finalGrid),
            status,
            _events.ToArray(),
            totalChanges,
            statistics);
    }

    private int CountChanges(string algorithm) =>
        _events.Where(item => item.Algorithm == algorithm).Sum(item => item.Changes.Count);
}
